src/zycast.c

```diff
@@ -0,0 +1,336 @@
+ *
+ *   4 byte signature: 'z', 'y', 'x', 0x0 [1]
+ *   16 bit checksum [2][3]
+ *   32 bit packet id [2][4]
+ *   32 bit packet length [2][5]
+ *   32 bit file length [2][6]
+ *   32 bit image bitmap [2][7]
+ *   2 byte ascii country code [8]
+ *   8 bit  flags [9]
+ *   5 byte reserved [10]
+ *
+ * [1] the terminating null is not actually checked by the observed
+ *     implemtations, but is assumed to be safest in case the
+ *     signature is treated as a string
+ *
+ * [2] all integers are in network byte order, i.e. big endian
+ *
+ * [3] checksum = sum >> 16 + sum, where sum is the sum of all
+ *     payload bytes
+ *
+ * [4] starts at 0 and is incremented by 1 for each packet.  Used both
+ *     to ensure sequential, loss free, unidirectional transport, and to
+ *     allow the transfer to start at any point.  The sequence must be
+ *     repeated until the transfer is complete
+ *
+ * [5] Testing indicates that some implementations expect 1024 byte
+ *     packets.  Smaller size results in a corrupt download, and larger
+ *     size causes the download to hang - waiting for packet ids which
+ *     does not exist.
+ *
+ * [6] the length of each file in case of a multi file transfer.
+ *
+ *
+ * [7] the lower 8 bits is a bitmap of all image types included in the
+ *     transfer.  Bits 8 - 16 contains the image type for this packet.
+ *     The purpose of the upper 16 bits is unknown.
+ *
+ *     The known image types are
+ *
+ *       0x01 - "bootbase" (often "Bootloader" partition)
+ *       0x02 - "rom"      (often "data" partition)
+ *       0x04 - "ras"      (often "Kernel" partition)
+ *       0x08 - "romd"     (often "rom-d" partition)
+ *       0x10 - "backup"   (often "Kernel2" partition)
+ *
+ *     The supported set of images vary among implementations.
+ *     The protocol may support other image types.
+ *
+ *     WARNING: The location of each supported image type is hard
+ *      coded in the bootloader server implementation.  There is no
+ *      relation to the bootloader configuration, and no way to verify
+ *      that those values are correct without decompiling that
+ *      implementations. Device specific bugs are likely, and may
+ *      result in a brick.
+ *
+ * [8] two upper case ascii characters, like 'D','E'. The purpose
+ *     is unknown, but ZyXEL devices are often configured with this
+ *     as one of their device specific variables
+
+ * [9] bitmap controlling actions taken after a complete transfer:
+ *
+ *       0x01 - set DebugFlag
+ *       0x02 - erase "rom"
+ *       0x04 - erase "rom-d"
+ *
+ *     Other, unknown, values may exist in the protocol.  Device
+ *     support may vary.
+ *
+ * [10] these bytes are not used by the observed implementations.
+ *      The purpose is therefore unknown. There is a risk
+ *      they are interpreted by other devices, resulting in
+ *      unexpected and potentionally harmful behaviour.
+ *
+ * Copyright (C) 2023 Bjørn Mork <bjorn@mork.no>
```

Comment:
I like the effort you made here! This tool is such a footgun... Just the perfect thing to add to the collection :-)